src/meta/src/stream/stream_graph.rs

```diff
@@ -184,359 +175,399 @@ impl StreamActorBuilder {
-            actors: downstream_actors,
-            same_worker_node,
-        });
+    pub fn add_dispatcher(&mut self, dispatcher: Dispatcher) {
+        self.downstreams
+            .try_insert(dispatcher.dispatcher_id, dispatcher)
+            .unwrap();
     }
 
-    /// Build an actor after seal.
-    pub fn build(&self) -> StreamActor {
-        let dispatcher = self
-            .downstreams
-            .iter()
-            .map(
-                |StreamActorDownstream {
-                     dispatch_strategy,
-                     dispatcher_id,
-                     actors,
-                     same_worker_node: _,
-                 }| Dispatcher {
-                    downstream_actor_id: actors.as_global_ids(),
-                    r#type: dispatch_strategy.r#type,
-                    column_indices: dispatch_strategy.column_indices.clone(),
-                    // will be filled later by stream manager
-                    hash_mapping: None,
-                    dispatcher_id: *dispatcher_id,
-                },
-            )
-            .collect_vec();
+    /// Add an upstream to this actor.
+    pub fn add_upstream(&mut self, upstream: StreamActorUpstream) {
+        self.upstreams
+            .try_insert(upstream.edge_id, upstream)
+            .unwrap();
+    }
 
-        StreamActor {
-            actor_id: self.actor_id.as_global_id(),
-            fragment_id: self.fragment_id.as_global_id(),
-            nodes: Some(self.nodes.deref().clone()),
-            dispatcher,
-            upstream_actor_id: self
-                .upstreams
-                .iter()
-                .flat_map(|(_, StreamActorUpstream { actors, .. })| actors.0.iter().copied())
-                .map(|x| x.as_global_id())
-                .collect(), // TODO: store each upstream separately
-            colocated_upstream_actor_id: if self.chain_same_worker_node {
-                if self.upstreams.is_empty() {
-                    None
-                } else {
-                    Some(ColocatedActorId {
-                        id: self
-                            .upstreams
-                            .values()
-                            .exactly_one()
-                            .unwrap()
-                            .actors
-                            .as_global_ids()
-                            .into_iter()
-                            .exactly_one()
-                            .unwrap(),
-                    })
+    /// Rewrite the actor body.
+    ///
+    /// During this process, the following things will be done:
+    /// 1. Replace the logical `Exchange` in node's input with `Merge`, which can be executed on the
+    /// compute nodes.
+    /// 2. Fill the upstream mview info of the `Merge` node under the `Chain` node.
+    fn rewrite(&self) -> MetaResult<StreamNode> {
+        self.rewrite_inner(&self.nodes, 0)
+    }
+
+    fn rewrite_inner(&self, stream_node: &StreamNode, depth: usize) -> MetaResult<StreamNode> {
```

Comment:
This is basically the same as the previous `ActorGraphBuilder::build_inner` which turns `Exchange` to `Merge`. However, it's obviously more like a per-actor stuff instead of an actor-graph stuff, so I move it to `ActorBuilder` and slightly refactor the logic.